execbuilder: add "average lookup ratio" parallelization heuristic for lookup joins

yuzefovich opened the pull request:
**sql: use correct context in recursive CTE iterations**

Previously, the `execbuilder.Builder` that we use in recursive CTE
iterations referenced the same context that the Builder captured on the
main query path. This is problematic since that context might have
a tracing span that's already been finished. This commit fixes this
issue by explicitly passing the context argument into the iteration
function. This was only exposed because we added some logging in the
following commit but has been present for a while now.

**logictest: extend a few EXPLAIN tests a bit**

This commit adjusts a few existing EXPLAIN tests to use VERBOSE as well as
adds another RBR setup with CASCADE FKs. These will be used to highlight
the parallelization change in the multi-key lookup joins in the
following commit.

**execbuilder: add "average lookup ratio" parallelization heuristic for lookup joins**

The DistSender API forces its users to make a choice between cross-range
parallelism (which is needed for performance) and setting memory limits
(which is needed for stability). Streamer was introduced to address this
limitation, but it comes with some requirements, one of which is that it
needs to have access to LeafTxns. However, mutation statements must run
in the RootTxn, so we never use the Streamer for those and fall back to
using the DistSender API directly (via `txnKVFetcher`). There, we
currently have the following heuristic:
- if we know that each input row results in at most one lookup row, then
we consider such a lookup to be "safe" parallelization, so we disable
usage of memory limits on the BatchHeader. This is the case for index
joins (when we always expect to get exactly one looked up row) as well
as lookup joins that have "equality columns are key" property (when we
expect at most one looked up row).
- otherwise, if we have a multi-key lookup join, we use the default
fetcher memory limits (TargetBytes of 10MiB), which disables cross-range
parallelism.

Most commonly this will affect mutation statements and will have a more
pronouanced effect on the multi-region tables, so this commit extends the
heuristic for when we consider it to be "safe" for parallelization.

Namely, we now calculate the average lookup ratio based on the lookup
equality columns and the available table / column statistics, and if the
ratio doesn't exceed the allowed limit, then we'll enable the
parallelism. To a certain degree, this heuristic resembles the "equality
columns are key" heuristic that we already utilize, but instead of
a guaranteed maximum on the lookup ratio we use the estimated average.

What we're trying to prevent with the existing and the new heuristics is
the case when we construct such a KV batch that the KV response will
overwhelm (read "will OOM") the node issuing the KV batch. In the
existing heuristic we say that "if lookup ratio is guaranteed to not
exceed one, then it should be safe".

I believe that the new heuristic should be safe in practice for most
deployments due to the following reasons:
- we already have an implicit limiting behavior in the join reader due
to its execution model (it first buffers some number of rows, up to
2MiB in size when not using the streamer, deduplicates the lookup spans,
and performs the lookup of all those spans in a single KV batch).
Empirical testing shows that we expect to have at most 25k lookups in
that single KV batch.
- this will have impact only when the streamer is not used, which most
commonly will mean we're executing a mutation, and in our docs we
advocate for not performing large mutations. (I'm stretching things
a bit here since even if we modify small amount of data, to compute that
we might read a lot, which could be destabilizing if we disable KV
limits. Yet a similar argument could be made that our current
"equality columns are key" heuristic is not safe - it's possible to
construct a scenario where we look up large amounts of data.)

In order to prevent this new heuristic from exploding in some edge
cases, two guardrails are added:
- in order to handle a scenario where the lookup ratio is not evenly
distributed (i.e. different input rows can result in vastly different
number of looked up rows), we'll disable the heuristic if the max lookup
ratio exceeds the allowed limit.
- in order to handle a scenario where looked rows are very large, we'll
disable the heuristic if the estimated average lookup row size exceeds
the allowed limit. (Note that we don't have this kind of protection in
the existing heuristics.)

I plan to do some more empirical runs to fine-tune the default values of
the newly added session variables, but the current defaults are:
- `parallelize_multi_key_lookup_joins_avg_lookup_ratio = 10`
- `parallelize_multi_key_lookup_joins_max_lookup_ratio = 10000`
- `parallelize_multi_key_lookup_joins_avg_lookup_row_size = 100 KiB`.

In order to de-risk rollout of this feature, we will initially apply the
new heuristic only to mutations of multi-region tables. New session
variable `parallelize_multi_key_lookup_joins_only_on_mr_mutations` can
be set to `false` to apply the heuristic to all statements, regardless
of the table being multi-region.

Fixes: #134351.
Epic: CRDB-44104

Release note (performance improvement): Mutation statements (UPDATEs and
DELETEs) that perform lookup joins into multi-region tables (perhaps as
part of a CASCADE) are now more likely to parallelize the lookups across
ranges which improves their performance.

yuzefovich committed:
execbuilder: add "average lookup ratio" parallelization heuristic for lookup joins

The DistSender API forces its users to make a choice between cross-range
parallelism (which is needed for performance) and setting memory limits
(which is needed for stability). Streamer was introduced to address this
limitation, but it comes with some requirements, one of which is that it
needs to have access to LeafTxns. However, mutation statements must run
in the RootTxn, so we never use the Streamer for those and fall back to
using the DistSender API directly (via `txnKVFetcher`). There, we
currently have the following heuristic:
- if we know that each input row results in at most one lookup row, then
we consider such a lookup to be "safe" parallelization, so we disable
usage of memory limits on the BatchHeader. This is the case for index
joins (when we always expect to get exactly one looked up row) as well
as lookup joins that have "equality columns are key" property (when we
expect at most one looked up row).
- otherwise, if we have a multi-key lookup join, we use the default
fetcher memory limits (TargetBytes of 10MiB), which disables cross-range
parallelism.

Most commonly this will affect mutation statements and will have a more
pronouanced effect on the multi-region tables, so this commit extends the
heuristic for when we consider it to be "safe" for parallelization.

Namely, we now calculate the average lookup ratio based on the lookup
equality columns and the available table / column statistics, and if the
ratio doesn't exceed the allowed limit, then we'll enable the
parallelism. To a certain degree, this heuristic resembles the "equality
columns are key" heuristic that we already utilize, but instead of
a guaranteed maximum on the lookup ratio we use the estimated average.

What we're trying to prevent with the existing and the new heuristics is
the case when we construct such a KV batch that the KV response will
overwhelm (read "will OOM") the node issuing the KV batch. In the
existing heuristic we say that "if lookup ratio is guaranteed to not
exceed one, then it should be safe".

I believe that the new heuristic should be safe in practice for most
deployments due to the following reasons:
- we already have an implicit limiting behavior in the join reader due
to its execution model (it first buffers some number of rows, up to
2MiB in size when not using the streamer, deduplicates the lookup spans,
and performs the lookup of all those spans in a single KV batch).
Empirical testing shows that we expect to have at most 25k lookups in
that single KV batch.
- this will have impact only when the streamer is not used, which most
commonly will mean we're executing a mutation, and in our docs we
advocate for not performing large mutations. (I'm stretching things
a bit here since even if we modify small amount of data, to compute that
we might read a lot, which could be destabilizing if we disable KV
limits. Yet a similar argument could be made that our current
"equality columns are key" heuristic is not safe - it's possible to
construct a scenario where we look up large amounts of data.)

In order to prevent this new heuristic from exploding in some edge
cases, two guardrails are added:
- in order to handle a scenario where the lookup ratio is not evenly
distributed (i.e. different input rows can result in vastly different
number of looked up rows), we'll disable the heuristic if the max lookup
ratio exceeds the allowed limit.
- in order to handle a scenario where looked rows are very large, we'll
disable the heuristic if the estimated average lookup row size exceeds
the allowed limit. (Note that we don't have this kind of protection in
the existing heuristics.)

I plan to do some more empirical runs to fine-tune the default values of
the newly added session variables, but the current defaults are:
- `parallelize_multi_key_lookup_joins_avg_lookup_ratio = 10`
- `parallelize_multi_key_lookup_joins_max_lookup_ratio = 10000`
- `parallelize_multi_key_lookup_joins_avg_lookup_row_size = 100 KiB`.

In order to de-risk rollout of this feature, we will initially apply the
new heuristic only to mutations of multi-region tables. New session
variable `parallelize_multi_key_lookup_joins_only_on_mr_mutations` can
be set to `false` to apply the heuristic to all statements, regardless
of the table being multi-region.

Release note (performance improvement): Mutation statements (UPDATEs and
DELETEs) that perform lookup joins into multi-region tables (perhaps as
part of a CASCADE) are now more likely to parallelize the lookups across
ranges which improves their performance.

yuzefovich committed:
logictest: extend a few EXPLAIN tests a bit

This commit adjusts a few existing EXPLAIN tests to use VERBOSE as well as
adds another RBR setup with CASCADE FKs. These will be used to highlight
the parallelization change in the multi-key lookup joins in the
following commit.

Release note: None

yuzefovich commented:
This is RFAL.

There are a few TODOs in the code that I'm curious to hear thoughts on. I also plan to do some more empirical testing to fine-tune the defaults but that doesn't need to block the review.

michae2 reviewed (commented):
The code looks great.

I'm wary of depending on stats for this, but I can see that there isn't really a good alternative. I'm worried that this is going to be "flaky" for customers, sometimes being used and sometimes not, leading to high P99 latencies and escalations where we try to figure out why parallelism isn't being used. Is there a way to make it more deterministic? Or at least to simplify it such that it's easy to understand when it will kick in and when it won't?

I'll keep thinking about it tonight. I think what you have is good. I just wonder if there's some small tweak to make it a little easier to understand for customers / more deterministic.

> Reviewed 7 of 7 files at r1, 15 of 20 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-:-OUb67OrCzUZAZ3qvUcY:b-ofzkbg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbCDd_Ab2lGiMK6oR8:-OUbCDd_Ab2lGiMK6oR9:bean3zv):*
It would be good to include descriptions of the session variables in the release note as well.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2790 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbMbO4AIY3ShgfJMTe:-OUbMbO4AIY3ShgfJMTf:bu1innp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2790)):*
> ```Go
> 	if allowedMaxRatio := sd.ParallelizeMultiKeyLookupJoinsMaxLookupRatio; allowedMaxRatio != 0 && colStat.Histogram != nil {
> 		// TODO: do we want to disable parallelization if we don't have a
> 		// histogram?
> ```

Joins with multiple equality columns will never have histograms. I think enabling parallelization is the right choice if there are no histograms.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbN4_c9HDnYFL_jYCW:-OUbN4_c9HDnYFL_jYCX:bn0o180) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2802)):*
> ```Go
> 	if allowedAvgRowSize := sd.ParallelizeMultiKeyLookupJoinsAvgLookupRowSize; allowedAvgRowSize != 0 && len(tableStats.AvgColSizes) > 0 {
> 		// TODO: do we want to disable parallelization if we don't have average
> 		// column sizes?
> ```

I think enabling parallelization is the right choice if we don't have average column size.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2780 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUfZ0ZD-3wrjM4HzX_u:-OUfZ0ZD-3wrjM4HzX_v:b7vs08q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2780)):*
> ```Go
> 	// NULLs in the equality columns.
> 	estimatedAvgRatio := tableStats.RowCount / colStat.DistinctCount
> 	if estimatedAvgRatio > allowedAvgRatio {
> ```

Instead of basing this heuristic on stats, WDYT of basing it on distribution? (Maybe something like `len(memo.GetLookupJoinLookupTableDistribution(...)) > 1`?)

I'm grasping for something that isn't stats. I'm really wary of basing this on stats. I think it will potentially cause more confusion than it is worth.

It's unfortunate that we can't use the streamer here. That would be the best solution.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2780 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUfZ0ZD-3wrjM4HzX_u:-OUfcGv63uM50y6ggQ_q:b51lo9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2780)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Instead of basing this heuristic on stats, WDYT of basing it on distribution? (Maybe something like `len(memo.GetLookupJoinLookupTableDistribution(...)) > 1`?)

I'm grasping for something that isn't stats. I'm really wary of basing this on stats. I think it will potentially cause more confusion than it is worth.

It's unfortunate that we can't use the streamer here. That would be the best solution.
</blockquote></details>

Or how about distribution + this is a mutation statement?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2780 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUfZ0ZD-3wrjM4HzX_u:-OUg8nZf4YxfOCVj4hxL:baynkth) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2780)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Or how about distribution + this is a mutation statement?
</blockquote></details>

Along the same lines: instead of adding these three additional session variables, maybe it would be simpler to add an "auto" value for `parallelize_multi_key_lookup_joins_enabled`, to better match `distsql` and `vectorize`? And maybe "auto" means a heuristic?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Just pushed some bug fixes after looking at the bundle from the customer:
- I made an incorrect assumption that we always constrain the prefix of index columns with equality conditions. This isn't necessarily the case, so now when we have `LookupExpr`, we explicitly track ColumnIDs of columns that were constrained via equalities.
- Turns out that we do want to pay attention to NULLs (both when estimating the average and the maximum lookup ratios). Rows that have a NULL value in one of the lookup columns cannot be looked up via lookup join (because no input row would match that NULL-lookup row, right?), so we need to subtract the null count from the table row count, as well as ignore the zeroth histogram bucket if it's a fake one we created for NULLs.
- I also bumped the default max ratio from 1k to 10k.

After applying these fixes I get the following behavior on the customer bundle:
- it has 35 lookup joins (out of which 18 have "equality cols are key" property), so we have 17 lookup joins to consider
- all 17 have the avg lookup ratio in 1-3 range
- 14 now get the parallelization (bumping the max allowed ratio included 3 more into this number, with 1k max ratio default we get 11)
- 2 have max ratio of 18k and 31k, so bumping the max value from 10k to 100k gets these
- the remaining one has max ratio of 365k, so it'd require 500k max ratio or so.

Based on these considerations, I picked 10k as seemingly reasonable default.

----

(Let's move the fundamental discussion to the top-level thread.)

> I'm wary of depending on stats for this, but I can see that there isn't really a good alternative. I'm worried that this is going to be "flaky" for customers, sometimes being used and sometimes not, leading to high P99 latencies and escalations where we try to figure out why parallelism isn't being used. Is there a way to make it more deterministic? Or at least to simplify it such that it's easy to understand when it will kick in and when it won't?

I share your concern about this being non-deterministic, and then whenever a new stats get collected, the execution behavior might change. This would increase the likelihood of tickets that we already get when due to new stats a new logical plan is chosen - we'd now have to consider the execution behavior too for this. I can't really think of a better signal to roll out this heuristic in a safe fashion.

> Instead of basing this heuristic on stats, WDYT of basing it on distribution? (Maybe something like `len(memo.GetLookupJoinLookupTableDistribution(...)) > 1`?)
> Or how about distribution + this is a mutation statement?

Can you shed more light on what "distribution" means in this context? Is this the "distribution" property that we have in the optimizer? IIUC it would give us a signal whether the data is spread across regions, and I'm worried that if the heuristic simply says "multi-region distribution -> parallelize", then it'd lead to OOMs since it'd have no guardrails. I share the same worry if the heuristic will be "multi-region enabled table -> parallelize".

Restricting the heuristic to apply only within a mutation seems like a good idea. We could also use the distribution or the table being MR-enabled as another condition for when we apply the heuristic, but I do think we need some kind of additional data-driven guardrail.

Maybe a better mechanism would be akin to query cache where we would run first N executions of the query without parallelization, and then once we built enough confidence that the amount of data scanned is low, then we'd enable parallelization on all consequent executions? This would also be somewhat non-deterministic, but at least it wouldn't alternate for query executions that happen one after another. (This seems complicated, so it'd be out of scope for 25.3.)

> It's unfortunate that we can't use the streamer here. That would be the best solution.

I agree that this is the ideal solution we should strive for. It just wasn't feasible in 25.3 time frame. (I do hope to polish some of the prototypes I had like #148415 to make it easier in the future if we choose to pursue this project.)

> I'm grasping for something that isn't stats. I'm really wary of basing this on stats. I think it will potentially cause more confusion than it is worth.

My hope is that the defaults we pick for the heuristic will be generous enough so that minor changes in the data would still either be always within allowed range, or always outside of the allowed range. If there is a significant change in the data, then yeah, we could definitely get unpredictable p99 latencies, or sporadic changes in the query latency. I don't have a good answer here :/

> Along the same lines: instead of adding these three additional session variables, maybe it would be simpler to add an "auto" value for `parallelize_multi_key_lookup_joins_enabled`, to better match `distsql` and `vectorize`? And maybe "auto" means a heuristic?

We could definitely expose a single session variable that would be the "parallelization mode", and it would allow values like "off", "auto" (current proposed defaults), perhaps we'd also add "conservative" and / or "aggressive" that would adjust the existing knobs accordingly. I think that this new heuristic we're working on is not something that we want to document (like we don't document the "equality cols are key" nor the 10k hard limit for scans when we parallelize), but I'd rather document the "parallelization mode" which would pick between several sets of defaults for the knobs, rather than those knobs directly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-:-OUgmknEA9uQigY26bet:b-k27jfd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`-- commits` line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbCDd_Ab2lGiMK6oR8:-OUgmG4H2yz_d5Li0He3:b-65iofi):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It would be good to include descriptions of the session variables in the release note as well.
</blockquote></details>

Let's hold off on this thought for now until we agree on the final shape of the heuristic. Ideally, I'd like to not document these session variables as they are not very user-friendly (and the power users who really want to use them should be able to read the comments in the proto files). Documenting them in the release notes kinda opens the variables for wider use than is currently intended.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2790 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbMbO4AIY3ShgfJMTe:-OUgmBRdDNi6ORGo9MTG:b-pbdqz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2790)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Joins with multiple equality columns will never have histograms. I think enabling parallelization is the right choice if there are no histograms.
</blockquote></details>

Agreed.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUbN4_c9HDnYFL_jYCW:-OUgm9zB2NBKuH9gGWG2:b-pbdqz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3edfe1d419bec062ffe95c546907f92d5dc8ab/pkg/sql/opt/exec/execbuilder/relational.go#L2802)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think enabling parallelization is the right choice if we don't have average column size.
</blockquote></details>

Agreed.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> I can't really think of a better signal to roll out this heuristic in a safe fashion.

I can't either, and it seems like this is the core of the problem. The proper way to ensure OOM-safety is to use some kind of flow control during execution, e.g. the streamer or admission control, not to guess ahead of time whether the BatchResponses will cause an OOM based on statistics.

The optimizer assumes that any operation it asks for is OOM-safe, and the execution engine provides this OOM-safety in various ways (spilling to disk, using distsql_workmem, etc). It's kind of an inversion for us to have to make a decision about OOM-safety in the optimizer. And it seems like it won't be very accurate.

> Can you shed more light on what "distribution" means in this context? Is this the "distribution" property that we have in the optimizer?

Yes, `memo.GetLookupJoinLookupTableDistribution` returns the optimizer distribution property of the lookup join.

> I'm worried that if the heuristic simply says "multi-region distribution -> parallelize", then it'd lead to OOMs since it'd have no guardrails. I share the same worry if the heuristic will be "multi-region enabled table -> parallelize".

Good point. I was trying to figure out if, by considering region, we could find another guarantee like we do for `LookupColsAreTableKey` but it seems like we cannot, since child tables could have any number of matching rows, even within a single region.

> My hope is that the defaults we pick for the heuristic will be generous enough so that minor changes in the data would still either be always within allowed range, or always outside of the allowed range. If there is a significant change in the data, then yeah, we could definitely get unpredictable p99 latencies, or sporadic changes in the query latency. I don't have a good answer here :/

I don't have a good answer, either. It seems like if we're wrong in either direction, there is a problem. If we under-estimate and pick parallelization then the node could OOM. If we over-estimate and don't pick parallelization then it might appear as high P99 latency which is hard to diagnose (and also hard to prevent, unless the application forces parallelization by setting `parallelize_multi_key_lookup_joins_enabled=on` for the query).

We're building plan management to give people more control over the optimizer, but plan management might not be able to influence this decision, either, since the decision is made in execbuild instead of during optimization. (Though I might be able to add something special for this decision.)

> We could definitely expose a single session variable that would be the "parallelization mode", and it would allow values like "off", "auto" (current proposed defaults), perhaps we'd also add "conservative" and / or "aggressive" that would adjust the existing knobs accordingly.

I like this UI. But the thought of turning on "auto" or "conservative" or "aggressive" for every workload by default makes me nervous. I think the failure modes are too risky to have it enabled everywhere, at least until we prove otherwise.

I'd be ok with shipping this disabled by default. WDYT?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-:-OUhplMq7URjOAd21mNr:bp9fept)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)



<!-- Sent from Reviewable.io -->

michae2 commented:
@rytaft talked me down a bit, and pointed out this is an innovation release, which is a great time to gather data about the efficacy of the heuristic. So I take it back about shipping it disabled. I'd be ok turning it on for some set of statements.

yuzefovich commented:
Pushed an update which makes the new heuristic only apply to mutations of multi-region tables by default (it can be applied to all stmts by setting new session variable `parallelize_multi_key_lookup_joins_only_on_mr_mutations` to `false`). Also reverted some of the test changes that were related to read-only stmts or mutations of non-multi-region tables.

yuzefovich committed:
sql: use correct context in recursive CTE iterations

Previously, the `execbuilder.Builder` that we use in recursive CTE
iterations referenced the same context that the Builder captured on the
main query path. This is problematic since that context might have
a tracing span that's already been finished. This commit fixes this
issue by explicitly passing the context argument into the iteration
function. This was only exposed because we added some logging in the
following commit but has been present for a while now.

Release note: None

yuzefovich commented:
Newly added logging exposed incorrect context capture during recursive CTE iterations ("span use after finish"), so I added another commit to fix that.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 22 files at r8, 21 of 21 files at r9, 4 of 4 files at r10, 17 of 17 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-:-OUv1ufK6gkuRN4c73co:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2766 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUuyyDc42L_N0hMBZ9M:-OUuyyDc42L_N0hMBZ9N:b5iuem8) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d661b98fa43bf48a9b298e29d1ecdffa629c4/pkg/sql/opt/exec/execbuilder/relational.go#L2766)):*
> ```Go
> 	// rows each input row will result in, and if it's relatively small, then
> 	// we'll consider the lookup join "safe" for parallelization.
> 	tableStats, ok := memo.GetTableStats(md, join.Table)
> ```

commentary: IIUC, by using the statistics for the entire table, this heuristic will always give the same result regardless of the constants used in the query. That could create a performance cliff in the sense that if a single frequent value in the table crosses the `parallelize_multi_key_lookup_joins_max_lookup_ratio` threshold it will cause all lookup joins on that column to switch from parallel to non-parallel (or vice-versa) regardless of whether they read that specific frequent value.

If we want these calculations to depend on the constants used in the query, then instead of `memo.GetTableStats` they should be based on the ColumnStatistic returned by `memo.RequestColStat` for the LookupJoinExpr (maybe compared with the ColumnStatistic for the lookup join input).

On the one hand, by basing the calculation on table stats regardless of query constants, it will be easy to tell if a lookup join will use parallelism, because it will always be the same answer (until new stats are collected). On the other hand, when new stats are collected, every query could switch at once, which might be how an OOM happens if they all switch to using parallelism at once.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the review and the discussion!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-:-OUv5MQlA3Caao79YagQ:b-8adz5y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2766 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/148186#-OUuyyDc42L_N0hMBZ9M:-OUv4LTCEEtN55OF52FZ:b-fl6x8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d661b98fa43bf48a9b298e29d1ecdffa629c4/pkg/sql/opt/exec/execbuilder/relational.go#L2766)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

commentary: IIUC, by using the statistics for the entire table, this heuristic will always give the same result regardless of the constants used in the query. That could create a performance cliff in the sense that if a single frequent value in the table crosses the `parallelize_multi_key_lookup_joins_max_lookup_ratio` threshold it will cause all lookup joins on that column to switch from parallel to non-parallel (or vice-versa) regardless of whether they read that specific frequent value.

If we want these calculations to depend on the constants used in the query, then instead of `memo.GetTableStats` they should be based on the ColumnStatistic returned by `memo.RequestColStat` for the LookupJoinExpr (maybe compared with the ColumnStatistic for the lookup join input).

On the one hand, by basing the calculation on table stats regardless of query constants, it will be easy to tell if a lookup join will use parallelism, because it will always be the same answer (until new stats are collected). On the other hand, when new stats are collected, every query could switch at once, which might be how an OOM happens if they all switch to using parallelism at once.
</blockquote></details>

Interesting, I haven't really considered this angle since I assumed we want to make this decision based on the full table. I do like the current approach more than if we based it on the column stat specific to the lookup join expr since I like that it eliminates a possible alternation between "parallelize" and "not parallelize" based on the constant values (which seems like it'd be another dimension for unpredictability, even more so that changing of stats over time).

I also think that if we happen to switch from "not parallelize" to "parallelize" based on the new table stat, it shouldn't be more likely to OOM as the "direction" of change between previous and new table stats are into the "safer" state (meaning either the average lookup ratio decreased, or the heavy hitter's frequency decreased). The exception to this thinking is we get unlucky with sampling, and we miss some heavy hitters.

Let's keep it this way and see what happens in production :smile: 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
It's pretty random, but for posterity the main commit in this PR results in a microbenchmark regression in pgwire:
```
args: benchdiff "--run=^BenchmarkWriteBinaryColumnarDecimal" "--old=9f84aa54^" "--new=9f84aa54" "./pkg/sql/pgwire" "-b"

name                           old time/op    new time/op    delta
WriteBinaryColumnarDecimal-24     299µs ± 1%     387µs ± 1%  +29.39%  (p=0.000 n=9+10)

name                           old alloc/op   new alloc/op   delta
WriteBinaryColumnarDecimal-24    49.1kB ± 0%    49.1kB ± 0%     ~     (p=0.246 n=10+9)

name                           old allocs/op  new allocs/op  delta
WriteBinaryColumnarDecimal-24     2.05k ± 0%     2.05k ± 0%     ~     (all equal)
```
